### PR TITLE
[FIX] website: keep table of contents title & menu entries sync

### PR DESCRIPTION
**Pull request by tdecaluwe**
The table of contents menu entries are generated automatically which
poses a problem in translation mode. The menu translation entries
are not be editable separately, but the users might be trying to.

This commit shows a notification when the user clicks on the menu
entries while in translation mode, explaining that they are generated
from the title entries.
It was initially intended to use a tooltip - but the amount of code
needed to display a tooltip without marking the DOM as modified is
needlessly complex.
Additionally, the translation widget updates the generated links of
table of content headers to keep them synchronized.
This could not be done by matching the translation id because those
are different when a style is applied to the header.

To avoid that styles applied on a plain text during translation were
also appearing in the navigation menu, an attempt at adding a span
around them to make sure that their translation was distinct from the
one inside the main content. But this led to the risk of losing
existing translations.
Because of this, and because that situation seems unlikely, any
remaining style in the navigation menu is instead stripped when the
table of content is started to maintain consistency with what is shown
during translation.

task-2752391

### CODE REVIEW

**Comment by bso-odoo** [comment]
Extracted problems found during this PR to 14.0: https://github.com/odoo/odoo/pull/98069

File: addons/website/static/src/js/menu/translate.js [hunk 1]
```diff
@@ -178,6 +178,36 @@ var TranslatePageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
+            }).on('show.bs.tooltip', ev => {
+                this.translator.wysiwyg.odooEditor.observerUnactive();
+            }).on('shown.bs.tooltip', ev => {
+                this.translator.wysiwyg.odooEditor.observerActive();
+            }).on('hide.bs.tooltip', ev => {
+                this.translator.wysiwyg.odooEditor.observerUnactive();
+            }).on('hidden.bs.tooltip', ev => {
+                $target.tooltip('dispose');
+                this.translator.wysiwyg.odooEditor.observerActive();
+            }).tooltip('show');
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/translate.js:205]: That looks too fragile, despite probably working correctly IMHO. Especially knowing how the editor behaves with those observer things.
It looks like an open door for issues (both because of the tooltip mechanism and the editor).

Wouldn't we be better with a toaster/notification on click on this element? Would be a bit less good UX (maybe not actually) but more straightforward code and less prone to future error.

Note that I see a simpler form of what is done here there: https://github.com/odoo/odoo/blame/f1f3fcef6cc471fc01da574da26712e643315da6/addons/web_editor/static/src/js/wysiwyg/wysiwyg.js#L1782-L1784
Doesn't seems enough tho, I asked: https://github.com/odoo/odoo/pull/71168/files#r1008253534

File: addons/website/static/src/js/menu/translate.js [hunk 2]
```diff
@@ -178,6 +178,36 @@ var TranslatePageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
             devicePreview: false,
         };
 
+        const showTip = ev => {
+            const $target = $(ev.target);
+            let message = _t('This translation is not editable.');
+            if (ev.target.closest('.s_table_of_content_navbar_wrap')) {
+                message = _t('Translate header in the text. Menu is generated automatically.');
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/translate.js:185]: ?
We are in generic code not related to table of content
Clicking on any `.o_not_editable [data-oe-translation-id]` will show this tooltip :thinking: ?

File: addons/website/static/src/js/menu/translate.js [hunk 3]
```diff
@@ -246,6 +276,19 @@ var TranslatePageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
+            processRecordsCallback(record, el) {
+                const tocMainEl = el.closest('.s_table_of_content_main');
+                const headerEl = el.closest('h1, h2');
+                if (!tocMainEl || !headerEl) {
+                    return;
+                }
+                const headerIndex = [...tocMainEl.querySelectorAll('h1, h2')].indexOf(headerEl);
+                const tocMenuEl = tocMainEl.closest('.s_table_of_content').querySelectorAll('.table_of_content_link > span')[headerIndex];
+                if (tocMenuEl.textContent !== headerEl.textContent) {
+                    tocMenuEl.textContent = headerEl.textContent;
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/translate.js:288]: We already have a way to sync titles in regular edit mode, can't we somehow reuse/share it in translate mode? (Didn't check so open question)

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks for the PR, sorry for late review.

@bso-odoo 

> The menu translation entries
should not be editable separately.
> 
> This commit prevents the menu entries from being translatable, and
shows a tooltip instead, explaining to the user that they are
generated from the title entries.

This is somehow not true, I am not sure if it was at the time of the commit or not, but the menu is and has always been in `.o_not_editable`.
It is possible that the editor was not properly applying the `contenteditable=false` at the time of the commit (there were issues related to that).

So, now (and as it was always supposed to be), one can't edit the menu, regardless if it is in edit mode or translate mode: you can't.

Adding a tooltip is cool IMHO, but why doing that only in translate mode? This could/should be in regular edit mode too :+1: (maybe because at the time of the commit one could edit it?)

Probably to be rebased and adapted as clearly some fixes in editor were made regarding contenteditable.

File: addons/website/static/src/js/menu/translate.js (hunk 1 above)
Comment (bso-odoo) [reply addons/website/static/src/js/menu/translate.js]: Checked with PDE, he is fine with a notification - it is therefore a notification now.

File: addons/website/static/src/js/menu/translate.js (hunk 2 above)
Comment (bso-odoo) [reply addons/website/static/src/js/menu/translate.js]: Yes, it will appear as a yellow block in the screen - and it feels strange if it cannot be modified without explanation.
But I adapted the condition to only consider "non `:o_editable`" elements: (in case there is a chain of on/off editable)
```js
for (const translationEl of $('.o_not_editable [data-oe-translation-id]').not(':o_editable')) {
```

File: addons/website/static/src/js/menu/translate.js (hunk 3 above)
Comment (bso-odoo) [reply addons/website/static/src/js/menu/translate.js:273]: The edit mode sync mechanism is a full regeneration of the table of content on any change seen by a `MutationObserver` inside the "content" part of the snippet - it even generates new ids. Here we want to show that only a very specific text has changed, so I do not think it can be reused.
Replacing the other mechanism by this one also does not sound great because this can only be used to update existing labels - it would not work for adding missing ones or removing some.

**Comment by rdeodoo** [comment]
> Adding a tooltip is cool IMHO, but why doing that only in translate mode? This could/should be in regular edit mode too +1 (maybe because at the time of the commit one could edit it?)

I don't think this was processed?

It's now better in translate mode indeed as when you click on the TOC title the notification explains you that you can't edit it and you should instead edit the content which will sync the title
But we have the exact same issue with the regular edit mode (not translate mode), which seems even more important, right?

**Comment by bso-odoo** [comment]
> > Adding a tooltip is cool IMHO, but why doing that only in translate mode? This could/should be in regular edit mode too +1 (maybe because at the time of the commit one could edit it?)
> 
> I don't think this was processed?
> 
> It's now better in translate mode indeed as when you click on the TOC title the notification explains you that you can't edit it and you should instead edit the content which will sync the title But we have the exact same issue with the regular edit mode (not translate mode), which seems even more important, right?

I just discussed a bit about this with the PO.
It feels a bit strange, because selecting the menu is useful to configure its options (background color, position...) and it is rather likely that this selection is achieved by clicking on the menu entries. So we rather not have a notification for the normal usage. We also thought about showing it if the keyboard is used after clicking there - but since the content is not editable, the cursor is actually in a totally different location. (Not sure if this is a feature or a bug.)
Because of this, I do not think I will be able to `preventDefault` and show a notification on keyboard events when the menu part is selected.
TL;DR PO would rather not have it unless you have another suggestion.

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 4]
```diff
@@ -1919,6 +1919,9 @@ var SnippetsMenu = Widget.extend({
         if ($snippet && !$snippet.is(':visible')) {
             return;
         }
+        if (this.options.enableTranslation) {
```
Comment (rdeodoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1922]: I guess it makes sense since the snippet options column is not shown in translate mode but I'd like to have @qsm-odoo opinion here as this is 15.0.

**Review by rdeodoo — COMMENTED** [review]
Seems fine functionally, waiting for @qsm-odoo output about the line I commented before reviewing this further

File: addons/web_editor/static/src/js/editor/snippets.editor.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1922]: @rdeodoo @bso-odoo It should not be necessary IMO. If this is called while in translation mode it probably means the caller is also called in translation mode while it should not. We should disable top-most level things instead of low level calls.

Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: @qsm-odoo I moved the condition to `wysiwig`'s `cleanForSave` so that it does not delegate to `snippetsMenu` when we work on translations.

**Comment by bso-odoo** [comment]
@qsm-odoo Resolved conflicts and retested.

**Comment by qsm-odoo** [comment]
@robodoo delegate=xO-Tx

**Comment by bso-odoo** [comment]
Removed the `web_editor` hack that avoided the blocks flicker on save, and created a new dedicated task for it (task-3224668).

**Review by xO-Tx — APPROVED** [review]
@robodoo r+

**Comment by xO-Tx** [comment]
@bso-odoo don't forget to check / adapt in 16.0+ since the code to sync nav items to TOC content is already implemented
Thanks :+1:

**Comment by xO-Tx** [comment]
@robodoo retry
